ksql-engine/src/main/java/io/confluent/ksql/function/udaf/sum/ListSumUdaf.java

```diff
@@ -0,0 +1,223 @@
+
+      @Override
+      public Integer undo(final List<Integer> valueToUndo, final Integer aggregateValue) {
+        if (valueToUndo == null) {
+          return aggregateValue;
+        }
+        return aggregateValue - sumList(valueToUndo);
+      }
+
+      private int sumList(final List<Integer> list) {
+        int sum = 0;
+        for (Integer v: list) {
+          if (v == null) {
+            continue;
+          }
+          sum += v;
+        }
+        return sum;
+      }
+    };
+  }
+
+  @UdafFactory(description = "sum long values in a list into a single long")
+  public static TableUdaf<List<Long>, Long, Long> sumLongList() {
+    return new TableUdaf<List<Long>, Long, Long>() {
+
+      @Override
+      public Long initialize() {
+        return 0L;
+      }
+
+      @Override
+      public Long aggregate(final List<Long> valueToAdd, final Long aggregateValue) {
+        if (valueToAdd == null) {
+          return aggregateValue;
+        }
+        return aggregateValue + sumList(valueToAdd);
+      }
+
+      @Override
+      public Long merge(final Long aggOne, final Long aggTwo) {
+        return aggOne + aggTwo;
+      }
+
+      @Override
+      public Long map(final Long agg) {
+        return agg;
+      }
+
+      @Override
+      public Long undo(final List<Long> valueToUndo, final Long aggregateValue) {
+        if (valueToUndo == null) {
+          return aggregateValue;
+        }
+        return aggregateValue - sumList(valueToUndo);
+      }
+
+      private long sumList(final List<Long> list) {
+        long sum = 0L;
+        for (Long v: list) {
+          if (v == null) {
+            continue;
+          }
+          sum += v;
+        }
+        return sum;
+      }
+    };
+  }
+
+  @UdafFactory(description = "sum bigdecimal values in a list into a single bigdecimal")
+  public static TableUdaf<List<BigDecimal>, BigDecimal, BigDecimal> sumDecimalList() {
+    return new TableUdaf<List<BigDecimal>, BigDecimal, BigDecimal>() {
+
+      @Override
+      public BigDecimal initialize() {
+        return BigDecimal.ZERO;
+      }
+
+      @Override
+      public BigDecimal aggregate(final List<BigDecimal> valueToAdd,
+                                  final BigDecimal aggregateValue) {
+        if (valueToAdd == null) {
+          return aggregateValue;
+        }
+        return aggregateValue.add(sumList(valueToAdd));
+      }
+
+      @Override
+      public BigDecimal merge(final BigDecimal aggOne, final BigDecimal aggTwo) {
+        return aggOne.add(aggTwo);
+      }
+
+      @Override
+      public BigDecimal map(final BigDecimal agg) {
+        return agg;
+      }
+
+      @Override
+      public BigDecimal undo(final List<BigDecimal> valueToUndo, final BigDecimal aggregateValue) {
+        if (valueToUndo == null) {
+          return aggregateValue;
+        }
+        return aggregateValue.subtract(sumList(valueToUndo));
+      }
+
+      private BigDecimal sumList(final List<BigDecimal> list) {
```

Comment:
I didn't do it the way you suggest, as this would create a new decimal in each iteration. This is too much wasted memory when we talk about Big Data, no? I know there s a loss in precision but I figured it's a fair tradeoff?

Reply:
If people are using `DECIMAL` it is because they want precision, if they don't they can use `DOUBLE`.  Think financial accounting.  It is not OK to make the trade off on behalf of the user.  It should be their decision, not ours.